File: nim/all-your-base/all_your_base.nim
import algorithm, sequtils

proc fromInput(inputDigits: openArray[int], inputBase: int): int =
  for digit in inputDigits:
    result = result * inputBase + digit

proc toOutput(value: int, outputBase: int): seq[int] =
  var remainder = value
  var multiplier = 1

  while remainder > 0:
    multiplier *= outputBase

    let value = remainder mod multiplier
    let digit = value div (multiplier div outputBase)

    result.add digit
    remainder.dec value

  result.reverse

  if result.len == 0:
    result.add(0)

proc convert*(inputDigits: openArray[int], inputBase: int, outputBase: int): seq[int] =
  if inputBase < 2 or outputBase < 2 or inputDigits.anyIt(it < 0 or it >= inputBase):
    raise newException(ValueError, "Invalid bases")

  inputDigits.fromInput(inputBase).toOutput(outputBase)

File: ruby/all-your-base/all_your_base.rb
module BaseConverter
  def self.convert(input_base, digits, output_base)
    raise ArgumentError if input_base < 2 || output_base < 2

    return [0] unless digits.any?(&:positive?)

    from_decimal(output_base, to_decimal(input_base, digits.drop_while(&:zero?)))
  end

  def self.to_decimal(from_base, digits)
    digits.reduce(0) do |acc, digit|
      raise ArgumentError if digit.negative? || digit >= from_base

      acc * from_base + digit
    end
  end

  def self.from_decimal(to_base, num)
    digits = []
    remainder = num
    multiplier = 1

    while remainder > 0
      multiplier *= to_base

      value = remainder % multiplier
      digit = value / (multiplier / to_base)

      digits << digit
      remainder -= value
    end

    digits.reverse
  end
end
